### PR TITLE
build: conditional external dependency for cluster-ui

### PR DESCRIPTION
**Pull request by koorosh**
"@cockroachlabs/crdb-protobuf-client" package should be defined as external
dependency for `cluster-ui` webpack configuration only in case it is built
for `db-console`. Otherwise, it `cluster-ui` is built for publishing to
NPM registry, it should include bundled protobufs.

Release note: none

---
This change is [Reviewable](https://reviewable.io/reviews/jordanlewis/cockroach/9)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -360,7 +360,13 @@ CLUSTER_UI_JS := pkg/ui/cluster-ui/dist/main.js
 
 .SECONDARY: $(CLUSTER_UI_JS)
 $(CLUSTER_UI_JS): $(shell find pkg/ui/workspaces/cluster-ui/src -type f | sed 's/ /\\ /g') pkg/ui/yarn.installed pkg/ui/workspaces/db-console/src/js/protos.d.ts | bin/.submodules-initialized
+	$(NODE_RUN) -C pkg/ui/workspaces/cluster-ui yarn build --db-console-target
```
Comment (koorosh) [inline Makefile:363]: Run yarn build with `--db-console-target` arg to indicate that protobufs should be added to externals inside of webpack build

File: Makefile [hunk 2]
```diff
@@ -360,7 +360,13 @@ CLUSTER_UI_JS := pkg/ui/cluster-ui/dist/main.js
...
+
+# package-cluster-ui target can be used to build cluster-ui package for publishing to NPM registry
+.PHONY: package-cluster-ui
+package-cluster-ui: $(shell find pkg/ui/workspaces/cluster-ui/src -type f | sed 's/ /\\ /g') pkg/ui/yarn.installed pkg/ui/workspaces/db-console/src/js/protos.d.ts
 	$(NODE_RUN) -C pkg/ui/workspaces/cluster-ui yarn build
```
Comment (koorosh) [inline Makefile:368]: `cluster-ui` package can be built without `--db-console-target` in case this package will be published to NPM registry.

**Comment by jordanlewis** [comment]
I see... but ideally we'd do the same thing for CC too. It is too slow to build the `cluster-ui` package. Is there a way we can provide crdb-protobuf-client inside of CC console as a transitive dependency, to avoid adding the large protobuf code into the cluster-ui bundle?

**Comment by koorosh** [comment]
idk, the problem with CC console is that it doesn't depend on a single version of `cluster-ui` but depends on multiple versions at the same time because users can create clusters with different versions of CRDB. 
With assumption that generated protobufs code is always backward compatible might allow externalize protobufs from `cluseter-ui` completely, but this will require to rethink how CC Console consumes protobufs and always includes its latest version.

**Comment by jordanlewis** [comment]
Oh right, I keep forgetting about that.

Is there a way that you can publish an NPM module that isn't just a single big bundle? That would be another way to solve this problem.

**Comment by koorosh** [comment]
What if we generate JS protobufs the same way as generate Go protos?
- JS protobufs are generated and committed as part of source code;
- It has to be regenerated only when .proto definitions are changed;
- It will allow to avoid long builds every time UI has to be bundled;

**Comment by koorosh** [comment]
> Is there a way that you can publish an NPM module that isn't just a single big bundle? That would be another way to solve this problem.

NPM module can contain as many files as needed. 

While ago, we published protobuf-client to NPM registry (https://www.npmjs.com/package/@cockroachlabs/crdb-protobuf-client) to provide it as a dependency to `cluster-ui` but it added additional complexity during development.

**Comment by jordanlewis** [comment]
> It will allow to avoid long builds every time UI has to be bundled;

Actually the reason that the builds are so slow is not generating the JS in the first place. It's *packaging* the generated javascript up into an NPM main.js bundle in cluster-ui.

I was able to verify this claim via this PR. Try doing the cluster-ui build before and after this PR (specifically the externals change). The time is 2x faster on my machine - 15s vs 30s.

This is why I am suggesting that we could stop bundling up the protobuf javascript file. But I am not sure about how to do that technically. You say we can have as many files as needed, but how would we indicate that in the webpack config?

**Pull request closed without merging**